bin/client/src/interop/consolidate.rs

```diff
@@ -27,65 +28,79 @@ where
     P: PreimageOracleClient + Send + Sync + Debug + Clone,
     H: HintWriterClient + Send + Sync + Debug + Clone,
 {
-    let provider = OracleInteropProvider::new(oracle.clone(), boot.clone());
-
     info!(target: "client_interop", "Deriving local-safe headers from prestate");
 
-    // Ensure that the pre-state is a transition state.
+    // Ensure that the pre-state is a transition state. It is invalid to pass a non-transition state
+    // to this function, as it will not have the required information to derive the local-safe
+    // headers for the next super root.
     let PreState::TransitionState(ref transition_state) = boot.agreed_pre_state else {
         return Err(FaultProofProgramError::StateTransitionFailed);
     };
 
-    let block_hashes = transition_state
+    // Collect the cross-safe output roots and local-safe block hashes from the transition state.
+    let transition_meta = transition_state
         .pending_progress
         .iter()
         .zip(transition_state.pre_state.output_roots.iter())
         .map(|(optimistic_block, pre_state)| (pre_state, optimistic_block.block_hash))
         .collect::<HashMap<_, _>>();
 
-    let mut headers = Vec::with_capacity(block_hashes.len());
+    let mut headers = HashMap::default();
     let mut l2_providers = HashMap::default();
-    for (pre, block_hash) in block_hashes {
-        // Fetch the safe head's block hash for the given L2 chain ID.
-        let safe_head_hash =
-            fetch_output_block_hash(oracle.as_ref(), pre.output_root, pre.chain_id).await?;
+    for (cross_safe_output, local_safe_block_hash) in transition_meta {
+        // Fetch the cross-safe head's block hash for the given L2 chain ID.
+        let cross_safe_head_hash = fetch_output_block_hash(
+            oracle.as_ref(),
+            cross_safe_output.output_root,
+            cross_safe_output.chain_id,
+        )
+        .await?;
+
+        // Fetch the rollup config for the given L2 chain ID.
+        let rollup_config = ROLLUP_CONFIGS
+            .get(&cross_safe_output.chain_id)
+            .or_else(|| boot.rollup_configs.get(&cross_safe_output.chain_id))
+            .ok_or(FaultProofProgramError::MissingRollupConfig(cross_safe_output.chain_id))?;
+
+        // Initialize the local provider for the current L2 chain.
+        let mut local_provider = OracleL2ChainProvider::new(
+            cross_safe_head_hash,
+            Arc::new(rollup_config.clone()),
```

Comment:
I wonder if we can figure out a better way to reference the value in `kona_registry::ROLLUP_CONFIGS` across threads without cloning. Often time, we don't need the owned value. Bit of a footgun though since in some cases where updates are applied like derivation pipeline, we do want to reference the rollup config (system config) that is updated.

Reply:
Yeah, this is something I've noted down that I'm also not a fan of. I'll look to express this in a ticket so that we can get around to cleaning it up.

That said, thankfully it's not a terribly expensive clone.